imap_processing/ultra/decom_ultra.py

```diff
@@ -0,0 +1,160 @@
+# Standard
+import logging
+
+import numpy as np
+
+# Installed
+import xarray as xr
+
+# Local
+from imap_processing import decom
+
+logging.basicConfig(level=logging.INFO)
+#TODO: This is a work in progress. Hard-coded values will be removed.
+def read_n_bits(binary: str, n: int, current_position: int):
+    """
+    Extract a specified number of bits from a binary string,
+    starting from a given position.
+
+    Parameters
+    ----------
+    binary : str
+        The binary string from which bits will be read.
+    n : int
+        Number of bits to read from the binary string.
+    current_position : int
+        The starting position in the binary string from which bits will be read.
+
+    Returns
+    -------
+    value : int
+        The integer representation of the read bits or None if the end of the
+        string is reached before reading 'n' bits.
+    current_position + n
+        - The updated position in the binary string after reading the bits.
+    """
+
+    # Ensure we don't read past the end
+    if current_position + n > len(binary):
+        return None, current_position
+
+    value = int(binary[current_position:current_position + n], 2)
+    return value, current_position + n
+
+def log_decompression(value: int) -> int:
+
+    """
+    Perform logarithmic decompression on a 16-bit integer.
+
+    Parameters
+    ----------
+    value : int
+        A 16-bit integer comprised of a 4-bit exponent followed by a 12-bit mantissa.
+
+    Returns
+    -------
+    int
+        The decompressed integer value.
+    """
+    # The exponent e, and mantissa, m are 4-bit and 12-bit unsigned integers
+    # respectively
+    e = value >> 12  # Extract the 4 most significant bits for the exponent
+    m = value & 0xFFF  # Extract the 12 least significant bits for the mantissa
+
+    if e == 0:
+        return m
+    else:
+        return (4096 + m) << (e - 1)
+
+
+def decompress_binary(binary: str) -> list:
+    """
+    Decompress a binary string based on block-width encoding and
+    logarithmic compression.
+
+    This function interprets a binary string where the first 'width_bit' bits
+    specifies the width of the following values. Each value is then extracted and
+    subjected to logarithmic decompression.
+
+    Parameters
+    ----------
+    binary : str
+        A binary string containing the compressed data.
+
+    Returns
+    -------
+    list
+        A list of decompressed values.
+    """
+    current_position = 0
+    decompressed_values = []
+    width_bit = 5  # The bit width that describes the width of data in the block
+
+    while current_position < len(binary):
+        # Read the width of the block
+        width, current_position = read_n_bits(binary, width_bit, current_position)
```

Comment:
I read in the `read_n_bits` doc string that it's an int value of read bits. Why is that int value is used as width? Just curious why.